### PR TITLE
Habit comments on refactored entity

### PR DESCRIPTION
**Pull request by df7734**
### **This PR includes:** 

- All mappings & services for Comments
- Likes for Comments
- Replies for Comments
- Rating & Achievement calculation

### **Out of scope:**

- Tagging
- Notifications

### CODE REVIEW

**Review by IliaWithHat — APPROVED** [review]

**Review by HelenSotnik — APPROVED** [review]

**Review by ToriForH — APPROVED** [review]

File: core/src/main/java/greencity/config/SecurityConfig.java [hunk 1]
```diff
@@ -338,7 +349,7 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
                 .hasAnyRole(USER, ADMIN, MODERATOR, UBS_EMPLOYEE)
                 .requestMatchers(HttpMethod.GET,
                     "/newsSubscriber",
-                    "/comments",
+                    COMMENTS,
                     "/comments/{id}",
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/config/SecurityConfig.java]: please use `COMMENTS` constant here too

File: core/src/main/java/greencity/config/SecurityConfig.java [hunk 2]
```diff
@@ -132,6 +134,12 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
                     "/habit/statistic/{habitId}",
                     "/habit/statistic/assign/{habitAssignId}",
                     "/habit/statistic/todayStatisticsForAllHabitItems",
+                    HABITS + "/comments/{id}",
+                    HABITS + "/comments/{parentCommentId}/replies/active",
+                    HABITS + "/{habitId}/comments/count",
+                    HABITS + "/comments/{parentCommentId}/replies/active/count",
+                    // HABITS + "/comments/active",
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/config/SecurityConfig.java]: please remove commented code

File: core/src/main/java/greencity/controller/HabitCommentController.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+     * @return comment to certain habit specified by commentId.
+     */
+    @Operation(summary = "Get comment by id")
+    @ResponseStatus(value = HttpStatus.OK)
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = HttpStatuses.OK,
+            content = @Content(schema = @Schema(implementation = AddCommentDtoResponse.class))),
+        @ApiResponse(responseCode = "400", description = HttpStatuses.BAD_REQUEST,
+            content = @Content(examples = @ExampleObject(HttpStatuses.BAD_REQUEST))),
+        @ApiResponse(responseCode = "401", description = HttpStatuses.UNAUTHORIZED,
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/HabitCommentController.java]: based on the configs this endpoint is available for all users (you use `permitAll()` for this endpoint)
Please check for all added endpoint if you use correct configs and update response codes

File: core/src/test/java/greencity/ModelUtils.java [hunk 4]
```diff
@@ -921,4 +924,29 @@ public static UserAsFriendDto getUserAsFriendDto() {
             .chatId(1L)
             .build();
     }
+
+    public static PageableDto<CommentDto> getPageableCommentDtos() {
+        List<CommentDto> commentDtos = new ArrayList<>();
+
+        for (int i = 0; i < 5; i++) {
```
Comment (RomaMocherniuk) [inline core/src/test/java/greencity/ModelUtils.java]: please use stream

File: service-api/src/main/java/greencity/dto/comment/AddCommentDtoRequest.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+import org.hibernate.validator.constraints.Length;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@EqualsAndHashCode
+public class AddCommentDtoRequest {
+    @NotBlank(message = "The text of comment can not be empty")
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/comment/AddCommentDtoRequest.java]: please update message  to follow `NotBlank` documentation
`The annotated element must not be null and must contain at least one non-whitespace character. Accepts CharSequence.`

File: service-api/src/main/java/greencity/dto/comment/AddCommentDtoResponse.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+public class AddCommentDtoResponse {
+    @NotNull
+    @Min(1)
+    private Long id;
+
+    @NotEmpty
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/comment/AddCommentDtoResponse.java]: `NotEmpty`  does not support `CommentAuthorDto`

Supported types are:

```
CharSequence (length of character sequence is evaluated)
Collection (collection size is evaluated)
Map (map size is evaluated)
Array (array length is evaluated)
```

File: service-api/src/main/java/greencity/dto/comment/CommentAuthorDto.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@Data
+public class CommentAuthorDto {
+    @NotEmpty
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/comment/CommentAuthorDto.java]: `NotEmpty` does not support `Long`

File: service-api/src/main/java/greencity/dto/comment/CommentDto.java [hunk 8]
```diff
@@ -1,12 +1,41 @@
 package greencity.dto.comment;
 
-import lombok.AllArgsConstructor;
-import lombok.Data;
-import lombok.NoArgsConstructor;
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotEmpty;
+import jakarta.validation.constraints.NotNull;
+import lombok.*;
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/comment/CommentDto.java]: please do not use `*`

File: service-api/src/main/java/greencity/dto/comment/CommentDto.java [hunk 9]
```diff
@@ -1,12 +1,41 @@
...
+@NoArgsConstructor
+@Builder
+@EqualsAndHashCode
+@ToString
 public class CommentDto {
+    @NotNull
+    @Min(1)
+    private Long id;
+
+    @NotEmpty
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/comment/CommentDto.java]: same here

File: service-api/src/main/java/greencity/dto/comment/CommentDto.java [hunk 10]
```diff
@@ -1,12 +1,41 @@
...
+    private LocalDateTime createdDate;
+
+    @NotEmpty
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/comment/CommentDto.java]: same here

File: service-api/src/main/java/greencity/dto/comment/CommentVO.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    private Long articleId;
+
+    @Size(min = 1, max = 8000)
+    private String text;
+
+    @CreatedDate
+    private LocalDateTime createdDate;
+
+    @NotEmpty
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/comment/CommentVO.java]: same here

File: service-api/src/main/java/greencity/dto/comment/CommentVO.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
+package greencity.dto.comment;
+
+import greencity.dto.user.UserVO;
+import jakarta.validation.constraints.NotEmpty;
+import jakarta.validation.constraints.Size;
+import lombok.*;
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/comment/CommentVO.java]: please do not use `*`

File: service-api/src/main/java/greencity/dto/comment/CommentVO.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.List;
+import java.util.Set;
+
+@Getter
+@Setter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+@EqualsAndHashCode
+@ToString
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/comment/CommentVO.java]: please use `@Data`

File: service/src/main/java/greencity/service/CommentServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+
+    /**
+     * Method to check correctness of article and define its type.
+     *
+     * @param articleType {@link ArticleType}.
+     * @param articleId   {@link Long} id of an article.
+     *
+     * @return article author {@link User}.
+     */
+    private User articleCheckIfExistsAndReturnAuthor(ArticleType articleType, Long articleId) {
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/CommentServiceImpl.java]: it breaks Single responsibility principle
And you have code duplication here

File: service/src/main/java/greencity/service/CommentServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+    public int countCommentsForHabit(Long habitId) {
+        Habit habit = habitRepo.findById(habitId)
+            .orElseThrow(() -> new NotFoundException(HABIT_NOT_FOUND_BY_ID + habitId));
+        return commentRepo.countNotDeletedCommentsByHabit(habit.getId());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/CommentServiceImpl.java:221]: Any reason you did not add this for all methods?
```
/**
     * {@inheritDoc}
     */
```

File: service/src/test/java/greencity/ModelUtils.java [hunk 16]
```diff
@@ -139,41 +143,7 @@
...
-import greencity.entity.Place;
-import greencity.entity.ShoppingListItem;
-import greencity.entity.Specification;
-import greencity.entity.Tag;
-import greencity.entity.User;
-import greencity.entity.UserAchievement;
-import greencity.entity.UserAction;
-import greencity.entity.UserShoppingListItem;
-import greencity.entity.VerifyEmail;
+import greencity.entity.*;
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/ModelUtils.java]: please do not use `*`

File: service/src/test/java/greencity/ModelUtils.java [hunk 17]
```diff
@@ -182,18 +152,7 @@
...
-import greencity.enums.HabitAssignStatus;
-import greencity.enums.HabitRate;
-import greencity.enums.NotificationType;
-import greencity.enums.PlaceStatus;
-import greencity.enums.ProjectName;
-import greencity.enums.Role;
-import greencity.enums.ShoppingListItemStatus;
-import greencity.enums.TagType;
-import greencity.enums.UserStatus;
+import greencity.enums.*;
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/ModelUtils.java]: same here

**Comment by RomaMocherniuk** [comment]
Please increase test coverage at least to 90%
image

**Pull request merged**